Add signal-based heartbeat option

laurynasgadl opened the pull request:
Related: https://github.com/php-amqplib/php-amqplib/pull/775

Add the option to use a signal-based heartbeat sender, that will keep the connection alive even during lengthy message processing.

The heartbeat will not interrupt blocking actions (`sleep()` is interruptable and will return the remaining sleep time as a result https://www.php.net/manual/en/function.sleep.php).

## Usage
```php
<?php
$connection = AMQPStreamConnection::create_connection([
    ['host' => HOST, 'port' => PORT, 'user' => USER, 'password' => PASS, 'vhost' => VHOST]
], ['heartbeat' => 4]);

$sender = new PCNTLHeartbeatSender($connection);
$sender->register();
...
```

lukebakken commented:
Hello, thank you for taking the time to contribute to this library.

> The heartbeat will not interrupt blocking actions

My guess is that work done on a message is typically not interrupt-able. Can you give an example that doesn't use `sleep` of work that can be done on a message while being interrupted?

laurynasgadl commented:
In my case, there’s often a need to read stream output during message processing. The stream is read in a basic loop in specific chunk sizes. The problem comes, when the stream writer takes a while to produce the output, or there is a large amount of data being streamed.

ramunasd commented:
Many thanks, @laurynasgadl for Your efforts and time. I really appreciate Your attempt to solve the missed heartbeat problem.
However, I don't think this PR is production-ready. It would be great to use another approach and make this feature non-intrusive. So here are key points:
* no additional connection parameters - heartbeat checker isn't connection property
* do not duplicate or replace last activity timestamp - You can get it directly from connection IO
* heartbeat is affected by socket reads as well, please take 
* heartbeat check might brake another frame - this PR mark activity after full data write so nobody prevents from writing heartbeat frame in the middle of another frame

What do You think? Would it be possible to fix these things?

laurynasgadl commented:
Thanks for the feedback. These seem doable, will try it out and let you know.

laurynasgadl commented:
@ramunasd these are the changes I've made based on your points, let me know what you think:
>- no additional connection parameters - heartbeat checker isn't connection property

- Made PCNTL Heartbeat sender settable via `set_heartbeat_sender` instead of it being a connection option;

>- do not duplicate or replace last activity timestamp - You can get it directly from connection IO
>- heartbeat is affected by socket reads as well, please take

- Added `getLastActivity` method in IO & made it reachable via connection;

>- heartbeat check might brake another frame - this PR mark activity after full data write so nobody prevents from writing heartbeat frame in the middle of another frame

- Added a `writing` variable in connection thats is compared to _false_ in the heartbeat sender as an additional condition to be able to send a beat & added `last_heartbeat` variable in IO which is updated right before `write_heartbeat` to prevent a frame-breaking edge-case when async heartbeat is called during the execution of `write_heartbeat`.

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -191,6 +202,7 @@ public function __construct(
         $this->heartbeat = $heartbeat;
         $this->connection_timeout = $connection_timeout;
         $this->channel_rpc_timeout = $channel_rpc_timeout;
+        $this->heartbeat_sender = new NullHeartbeatSender();
```
Heartbeat sender isn't connection property. I would prefer something like that:
```
$heartbeatSender = new HeartbeatSender($connection);
$heartbeatSender->register();
$heartbeatSender->unregister();
```

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -262,6 +274,7 @@ protected function connect()
                 if (!$host) {
                     //Reconnected
                     $this->io->reenableHeartbeat();
+                    $this->heartbeat_sender->setHeartbeat($this->heartbeat);
```
please remove

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -278,6 +291,7 @@ protected function connect()
             $this->closeChannels();
             $this->close_input();
             $this->close_socket();
+            $this->heartbeat_sender->shutdown();
```
please remove

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -395,6 +412,7 @@ protected function do_close()
         $this->setIsConnected(false);
         $this->close_input();
         $this->close_socket();
+        $this->heartbeat_sender->shutdown();
```
please remove

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -700,6 +718,7 @@ public function close($reply_code = 0, $reply_text = '', $method_sig = array(0,
     {
         $result = null;
         $this->io->disableHeartbeat();
+        $this->heartbeat_sender->shutdown();
```
please remove

ramunasd commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -1028,6 +1064,15 @@ protected function setIsConnected($is_connected)
+    /**
+     * @param HeartbeatSenderInterface $sender
+     */
+    public function set_heartbeat_sender($sender)
```
please remove

ramunasd commented on PhpAmqpLib/Wire/IO/AbstractIO.php:
```diff
@@ -40,6 +40,9 @@ abstract class AbstractIO
     protected $last_write;
 
+    /** @var int|float */
+    protected $last_heartbeat;
```
last heartbeat will be same as last write?

laurynasgadl replied:
last_write is updated after the writing is done, while last_heartbeat will be updated before writing. This is needed due to the posibility of interrupt happening during `write_heartbeat,` when last_write has not been updated yet.

ramunasd replied:
Isn't it better to use `$conection->writing` to prevent whole heartbeat check comming from outside?

laurynasgadl replied:
`check_heartbeat` is called from IO::read as well, when `$connection->writing` is set to _false_

Instead of using `last_heartbeat`, I could add a `writing_heartbeat` state variable in IO and call `isWritingHeartbeat` from connection during `isWriting` as an additional condition. Not sure if that is a better solution, though

ramunasd replied:
Well yes, but IO::read is internal call before read, is not possible to interupt writeable frame there.
You add `$this->connection->isWriting()` protection to PCNTLHeartbeatSender, but in fact this protection is required for all external calls, i.e. tick based heartbeat checkers.
Also, time difference between last_write and last_heartbeat will be insignificant, hardly it will make any real difference for heartbeats longer than a few seconds.

ramunasd commented on PhpAmqpLib/Connection/Heartbeat/PCNTLHeartbeatSender.php:
```diff
@@ -0,0 +1,86 @@
+    private function registerListener($interval)
+    {
+        pcntl_signal(SIGALRM, function () use ($interval) {
+            if ($this->shutdown || $this->connection->isWriting()) {
```
please check whether `$this->connection->isConnected()`

ramunasd replied:
if `$this->shutdown` then we can unset `$this->connection` and unregister signal listener

ramunasd commented on tests/Functional/SignalHeartbeatTest.php:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    /**
+     * @test
```
please add `@group signals` and `@covers ...` with code parts You are testing

ramunasd commented:
Good work so far. We are really to close to ready feature. Let's fix tests and we will be ready to merge.

laurynasgadl commented:
`PCNTLHeartbeatSender` is currently placed in `PhpAmqpLib\Connection\Heartbeat`, now that this folder only contains a single file, it would, probably, be better to place it somewhere else.

ramunasd commented:
> `PCNTLHeartbeatSender` is currently placed in `PhpAmqpLib\Connection\Heartbeat`, now that this folder only contains a single file, it would, probably, be better to place it somewhere else.

We can leave it there, later I will add tick based heartbeat checker [1]. It's alternative for those who can't/don't want to use *nix signals.

[1] https://www.php.net/manual/en/function.register-tick-function.php

lukebakken commented:
Thanks!

RackovychV commented:
It doesn't work for me when publish message with await of confirms - AMQPTimeoutException occurs.

ramunasd commented:
@RackovychV please submit a new issue with all related info - the way how to reproduce issue, all related info, debug log.
